### PR TITLE
[CBRD-24460] Add 'docker for ctp'

### PR DESCRIPTION
**Pull request by swi0110**
Refer to http://jira.cubrid.org/browse/CBRD-24460

CTP(= CUBRID Test Program)
CTP's goal is test to the CUBRID project.
'docker for ctp' provides the ctp test env.

### CODE REVIEW

**Comment by hgryoo** [comment]
Please add `docker4ctp` in the Project Lists section of README file in the root folder.

File: docker_for_ctp/ctp_config/ha_repl.conf [hunk 1]
```diff
@@ -0,0 +1,83 @@
+# Copyright (c) 2016, Search Solution Corporation. All rights reserved.
```
Comment (hgryoo) [inline docker_for_ctp/ctp_config/ha_repl.conf:1]: please modify the license headers for all `*.conf` files

Comment (swi0110) [reply docker_for_ctp/ctp_config/ha_repl.conf:1]: Thank you.
I will reflect on it.

**Comment by hgryoo** [comment]
@memnoth If you have time, any comment would be helpful about the docker environment.

**Comment by hgryoo** [comment]
@swi0110 As far as I know, JDK 1.8 is required for CUBRID latest version. JDK 1.6 is installed in the change.

File: docker_for_ctp/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+        -e 's~^#PermitRootLogin yes~PermitRootLogin yes~g' \
+        /etc/ssh/sshd_config
+
+CMD service sshd restart
+EXPOSE 22
+
+# Need package for test
+RUN yum install -y net-tools sudo gdb
+RUN yum install -y git wget tar java-1.6.0-openjdk.x86_64 java-1.6.0-openjdk-devel.x86_64 dos2unix
+RUN yum install -y gcc-4.4.7 gcc-c++ libtool-2.2.6 cmake make
```
Comment (hgryoo) [inline docker_for_ctp/Dockerfile:21]: Are gcc-4.4.7 and gcc-c++ for running CTP? CUBRID Engine uses devtoolset-8-gcc and devtoolset-8-g++. Those versions are GCC 8.3.1

**Comment by swi0110** [comment]
@hgryoo
In the meantime, the CTP test was used to JDK 1.6.
because compatible on below the 10.x.
so JDK 1.8 is not verified to the CTP.

but If you want to version it up then I will change it.

File: docker_for_ctp/Dockerfile (hunk 2 above)
Comment (swi0110) [reply docker_for_ctp/Dockerfile:21]: The gcc-4.4.7 is the default version CentOS 7 as far as I know.
and It is used to CCI driver test.

If I version up to gcc then I can't guarantee to compatible the CCI drivier test cases.

**Comment by swi0110** [comment]
When I tested by JDBC unit test on JDK 1.8 then It was no problem.
So I up to the version by JDK 1.8.

File: docker_for_ctp/Dockerfile (hunk 2 above)
Comment (hgryoo) [reply docker_for_ctp/Dockerfile:21]: I see :-)

**Comment by hgryoo** [comment]
Where can I install docker? I think we need to add a requirements section.

File: docker_for_ctp/ctp_config/ctp_env_export.sh [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# !/bin/bash
+
+echo "export JAVA_HOME=/usr/lib/jvm/java-1.6.0-openjdk-1.6.0.41.x86_64" >> ${HOME}/.bash_profile
```
Comment (hgryoo) [inline docker_for_ctp/ctp_config/ctp_env_export.sh:3]: JAVA_HOME also needs to be changed

Comment (swi0110) [reply docker_for_ctp/ctp_config/ctp_env_export.sh:3]: thank you

**Comment by swi0110** [comment]
@hgryoo 
I add the requirements section.
please check by README.

**Comment by hgryoo** [comment]
I found a problem that connecting by ssh does not work. It seems that the docker network is not configured well.

**Comment by swi0110** [comment]
@hgryoo 
I checked your problem.
I think that problem is related to docker-decktop by windows.
When I installed by the docker-desktop in Windows, I met the occurred the same as your problem.
but, When I installed by docker in WSL, There is no problem.

I will add to this content in README.

File: docker_for_ctp/README.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
+# **Docker for CTP**
+## Supported Platforms
+- CentOS 7
+
+## Prerequisites
+You need to install Docker Engine on your host system. <br/>
+Please refer to the link, [Install Docker Engine](https://docs.docker.com/engine/install) to install.
+
+※ If you executed by Windows, please used WSL.
+(Refer to https://github.com/CUBRID/cubrid-contrib/pull/3#issuecomment-1223908324)
```
Comment (hgryoo) [inline docker_for_ctp/README.md:10]:
```suggestion
As a known issue, the docker network is not well configured when you install Docker on Windows.
Installing Ubuntu on WSL2 on Windows 10 is recommended.
Please refer to [the details](https://github.com/CUBRID/cubrid-contrib/pull/3#issuecomment-1223908324)
```

File: docker_for_ctp/README.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
+# **Docker for CTP**
+## Supported Platforms
+- CentOS 7
+
```
Comment (hgryoo) [inline docker_for_ctp/README.md:4]: It depends on the OS requirements of docker. we can remove this line.

File: docker_for_ctp/README.md [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+$ git clone https://github.com/CUBRID/cubrid-contrib.git
+$ cd cubrid-contrib/docker_for_ctp
+$ sh make_for_container.sh
+$ ssh ctp@172.200.1.2
+$ ctp.sh [test scenario] -c [ctp config file]
+```
+
+## What is the CTP?
+CTP(= CUBRID Test Progam)<br/>
+The goal is the test by CUBRID.<br/><br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:23]: How about describing CTP as follow? 
```suggestion
The goal of this tool is to execute test cases already written in [cubrid-testcases](https://github.com/CUBRID/cubrid-testcases) or test cases written by the user and report the results. For more details, please refer to [the link](https://github.com/CUBRID/cubrid-testtools/tree/develop/CTP)
```

File: docker_for_ctp/README.md [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+```shell
+$ git clone https://github.com/CUBRID/cubrid-contrib.git
+$ cd cubrid-contrib/docker_for_ctp
+$ sh make_for_container.sh
+$ ssh ctp@172.200.1.2
+$ ctp.sh [test scenario] -c [ctp config file]
+```
+
+## What is the CTP?
+CTP(= CUBRID Test Progam)<br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:22]: <br/> is not required. you can use 2 spaces.
```suggestion
CTP(= CUBRID Test Progam)  
```

File: docker_for_ctp/README.md [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+The goal is the test by CUBRID.<br/><br/>
+
+### How to use ctp?
+ctp.sh [Test scenario] -c [CTP/conf/*.conf]
+(ex : ctp.sh sql -c ~/CTP/conf/sql.conf )
+
+If you want to change the 'test scenario path', open the 'CTP/conf/*.conf'.<br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:29]:
```suggestion
If you want to change the 'test scenario path', open the 'CTP/conf/*.conf'.  
```

File: docker_for_ctp/README.md [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+and change the 'scenario' variable.<br/><br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:30]:
```suggestion
and change the 'scenario' variable.  
```

File: docker_for_ctp/README.md [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+The detail is followed URL<br/>
+https://github.com/CUBRID/cubrid-testtools <br/><br/>
+
+### How to create the container?
+Execute the 'sh make_for_container.sh'.<br/>
+It makes the 2 containers.<br/><br/>
+
+The container name is 'docker_for_ctp_01', 'docker_for_ctp_02'.<br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:39]:
```suggestion
Two containers are created with the names 'docker_for_ctp_01' and 'docker_for_ctp_02' respectively.  
```

File: docker_for_ctp/README.md [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+And you can connect the container by SSH.<br/><br/>
+
+### Container construct
+All user default password is 1234.<br/>
```
Comment (hgryoo) [inline docker_for_ctp/README.md:43]:
```suggestion
The default user name and password for the containers will be set as 'ctp' and '1234' respectively.  
```

File: docker_for_ctp/README.md [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+   - please execute 'sh ctp_config/ctp_env_export.sh'
+2. Where do I put the test build?
+   - ctp_config/test_build/
+   - 'ctp_config' dir is shared dir by the parent machine.
+3. Created core file during the test
+   - You can use the GDB command in the container.
+4. Can not CCI driver test
+   - please try to this
+      - cp $CUBRID/cci/include/* $CUBRID/include/
+      - cp $CUBRID/cci/lib/libcascci.so* lib/
```
Comment (hgryoo) [inline docker_for_ctp/README.md:68]: For copy & paste
```suggestion
      ```
      cp $CUBRID/cci/include/* $CUBRID/include/
      cp $CUBRID/cci/lib/libcascci.so* lib/
      ```
```

File: docker_for_ctp/README.md [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+### How to install the CUBRID?
+https://www.cubrid.org/manual/en/11.2/install.html#installing-and-running-cubrid-on-linux <br/><br/>
+
+### Q&A
+1. If can not work 'ctp.sh'
+   - please execute 'sh ctp_config/ctp_env_export.sh'
+2. Where do I put the test build?
+   - ctp_config/test_build/
+   - 'ctp_config' dir is shared dir by the parent machine.
+3. Created core file during the test
```
Comment (hgryoo) [inline docker_for_ctp/README.md:63]: Maybe we need to guide where the core files are created.
```suggestion
3. How to debug a core file that was created during the test?
```

**Review by hgryoo — APPROVED** [review]

File: docker_for_ctp/make_for_container.sh [hunk 14]
```diff
@@ -0,0 +1,73 @@
+#!/bin/bash
+
+dir=`pwd`
+VERSION=1.0
+IMAGE_NAME=ctp_container
```
Comment (memnoth) [inline docker_for_ctp/make_for_container.sh:5]: IMO it better has prefix such as `cubrid` to distinguish between this project and other containers, for instance,

```sh
PREFIX=cubrid
IMAGE_NAME="${PREFIX}_ctp_container"
```

`CONTAINER_NAME` as well.

**Review by memnoth — COMMENTED** [review]
@swi0110 

The changes should have better the following parts,

1. Just in case what happens on the following scenario that users restart their host machines? Should they run `make_for_container.sh` everytime they reboot? if so, does it remove all containers and images currently existing and re-install everything from beginning? if so, I don't think it is a good idea. It's gonna take kinda long time to prepare test envs.

\cc @hgryoo

File: docker_for_ctp/make_for_container.sh (hunk 14 above)
Comment (memnoth) [reply docker_for_ctp/make_for_container.sh:5]: It's up to you @swi0110 whether you take mind feedback or not. feel free :-)

Comment (swi0110) [reply docker_for_ctp/make_for_container.sh:5]: Thanks for your opinion, I will reflect on it.

**Comment by swi0110** [comment]
@memnoth 
Sorry, I don't understand your comment.
I think when the host machine reboot then you can start the container too.

**Comment by hgryoo** [comment]
> @memnoth Sorry, I don't understand your comment. I think when the host machine reboot then you can start the container too.

1) To persist data such as testing results, core files, etc, we need to provide a way to create a volume or mount it on the host machine.
2) The "Quick Start" in README may confuse the user that the make_for_container.sh file has to be executed every time it is used. We might need a more detailed explanation of container creation and usage.

**Comment by memnoth** [comment]
@swi0110 

Let's assume that I run `make_for_container.sh` at the first time and finished some CTP tests through ssh. Then I reboot my own host machine, after that, what should I do for doing another CTP test? like... can I connect to CTP container via ssh without any docker commands or manually launch new containers from the images?

Also @hgryoo mentioned good points that first, how could we get the result from containers. On the second, README should have more details such as what to do in that case.

**Comment by swi0110** [comment]
@hgryoo @memnoth 
I applied your comment.
but the volume mount on the host machine needs to think about it.
because CTP has many test result paths.

**Comment by hgryoo** [comment]
> @hgryoo @memnoth I applied your comment. but the volume mount on the host machine needs to think about it. because CTP has many test result paths.

I understand that the result files are created along with case files. How about cloning the `cubrid-testcases` repository out of the docker? In the make_for_container.sh, clone the `cubrid-testcases` in a directory of the host machine and mount it.

It makes sense because even if we do the HA replication test, both containers don't have to have the test cases.

**Comment by swi0110** [comment]
@hgryoo 
Sorry for the late.
I changed the container's design.

I added the 'test_repo' dir.
This dir is shared with all containers & parent machines.
And CTP & cubrid-testcases is located in this.
So the user can add or modify the test cases regardless of container status.

**Comment by memnoth** [comment]
LGTM

Yet there might be some issues we couldn't figure out but I don't think they are important at this time. At least the current version looks working. Anyway, you may still have ideas to go for more progress.

Here are good examples you can refer to,
- https://github.com/containers/toolbox
- https://github.com/89luca89/distrobox

Thank you for the contributions and good progress.

Best regards,
Leesoo

**Review by hgryoo — APPROVED** [review]

**Review by memnoth — APPROVED** [review]

**Comment by memnoth** [comment]
@mhoh3963

Could you review this PR as you are in reviewer please? or let me take it instead of you?

\cc @hgryoo

**Review by mhoh3963 — APPROVED** [review]

**Comment by swi0110** [comment]
Thanks for all reviews.
finally, I added to 'CTP & test cases dir located' on README.md.

**Pull request merged**